uplink/config.go

```diff
@@ -40,10 +43,58 @@ type RSConfig struct {
 // EncryptionConfig is a configuration struct that keeps details about
 // encrypting segments
 type EncryptionConfig struct {
-	Key       string      `help:"root key for encrypting the data"`
-	BlockSize memory.Size `help:"size (in bytes) of encrypted blocks" default:"1KiB"`
-	DataType  int         `help:"Type of encryption to use for content and metadata (1=AES-GCM, 2=SecretBox)" default:"1"`
-	PathType  int         `help:"Type of encryption to use for paths (0=Unencrypted, 1=AES-GCM, 2=SecretBox)" default:"1"`
+	// TODO: WIP#if/v3-1541 read TODO in Key method why the following field cannot
+	// be added
+	// RootKey     *storj.Key  // The root key for encrypting the data and read it from KeyFilepath, see Key method.
+	KeyFilepath string      `help:"the path to the file which contains the root key for encrypting the data"`
+	BlockSize   memory.Size `help:"size (in bytes) of encrypted blocks" default:"1KiB"`
+	DataType    int         `help:"Type of encryption to use for content and metadata (1=AES-GCM, 2=SecretBox)" default:"1"`
+	PathType    int         `help:"Type of encryption to use for paths (0=Unencrypted, 1=AES-GCM, 2=SecretBox)" default:"1"`
+}
+
+// Key returns the root key for encrypting the data which is stored in the path
+// indicated by KeyFilepath field.
+//
+// It returns an error if:
+//
+// * The file doesn't exist.
+// * The file doesn't have exactly 0400 permissions.
+// * There is an I/O error.
+//
+// TODO: WIP#if/v3-1541 refactor cache the key after the firs read. The problem
+// is with pkg/cfgstruct which is inflexible on accepting unexported keys nor
+// several different types as a type of storj.Key, [32]byte, etc.
+func (encCfg *EncryptionConfig) Key() (storj.Key, error) {
+	if encCfg.KeyFilepath == "" {
+		return storj.Key{}, nil
+	}
+
+	f, err := os.Open(encCfg.KeyFilepath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return storj.Key{}, Error.New("not found key file %q", encCfg.KeyFilepath)
+		}
+
+		return storj.Key{}, errs.Wrap(err)
+	}
+
+	fi, err := f.Stat()
+	if err != nil {
+		return storj.Key{}, errs.Wrap(err)
+	}
+
+	if p := fi.Mode().Perm(); p != os.FileMode(0400) {
```

Comment:
Why does it need to check the permissions at all? This probably will completely fail on file systems that don't have permissions.

Reply:
I don't know how this behaves in those platforms.
I was asking in the issue for the conventions, but I didn't get an answer, so I wanted to mimmic what `ssh` requires.

Do we remove it?

Reply:
I would remove it. I can easily see people making a portable setup and copying their key to a FAT32 usb drive.

Reply:
when we filed the ticket the point was definitely to encourage people to keep their secrets private. would it acceptable @egonelbre to require permissions on linux and osx builds? i don't think people should be using their key off of a usb key directly.

Reply:
mimicking the behavior of ssh was the intention for sure

Reply:
Meanwhile I see that we could say that we don't support FAT32 because "who is still using that? 😄" the point is Windows.

Unfortunately, I don't have an accessible Windows machine for testing it but the [`os.Chmod`](https://golang.org/pkg/os/#Chmod) documentation looks quite clear about it, something that I missed and @kaloyan-raev brought it up in this review.

Currently, I removed the permission thinks in my local, let me know if we want to keep it, but I think that we could skip it for now, because previous to this branch we have the passphrase in the same configuration file and this is one step to secure it, hence we can delay such thing for a future enhancement and have some discussion to identify tradeoffs and agree on in which ones we can afford.